modules/peridynamics/src/kernels/HorizonStabilizedFormIIFiniteStrainMechanicsNOSPD.C

```diff
@@ -0,0 +1,503 @@
+              dFdUk(_component, i) =
+                  _horizon_radius[nd] / origin_vec_nb2.norm() * origin_vec_nb2(i) * vol_nb2;
+
+            dFdUk *= _shape2[nd].inverse();
+
+            // calculate dJ/du
+            dJdU = 0.0;
+            for (unsigned int i = 0; i < 3; ++i)
+              for (unsigned int j = 0; j < 3; ++j)
+                dJdU += detF * invF(j, i) * dFdUk(i, j);
+
+            // calculate dS/du
+            dSdU = dSdFhat * dFdUk * _dgrad_old[nd].inverse();
+
+            // calculate dinv(F)Tdu
+            dinvFTdU.zero();
+            for (unsigned int i = 0; i < 3; ++i)
+              for (unsigned int J = 0; J < 3; ++J)
+                for (unsigned int k = 0; k < 3; ++k)
+                  for (unsigned int L = 0; L < 3; ++L)
+                    dinvFTdU(i, J) += -invF(J, k) * invF(L, i) * dFdUk(k, L);
+
+            // calculate the derivative of first Piola-Kirchhoff stress w.r.t displacements
+            dPxdUkx = dJdU * _stress[nd] * invF.transpose() + detF * dSdU * invF.transpose() +
+                      detF * _stress[nd] * dinvFTdU;
+
+            for (unsigned int i = 0; i < _nnodes; ++i)
+              for (unsigned int j = 0; j < _nnodes; ++j)
+                _local_ke(i, j) = (i == 0 ? -1 : 1) * (j == 0 ? 0 : 1) * _multi[nd] *
+                                  _horizon_radius[nd] / origin_vec_nb1.norm() *
+                                  (dPxdUkx * _shape2[nd].inverse()).row(_component) *
+                                  origin_vec_nb1 * _bond_status;
+
+            _assembly.cacheJacobianBlock(_local_ke, ivardofs, jvardofs, _var.scalingFactor());
+
+            if (_has_diag_save_in)
+            {
+              unsigned int rows = _nnodes;
+              DenseVector<Real> diag(rows);
+              for (unsigned int i = 0; i < rows; ++i)
+                diag(i) = _local_ke(i, i);
+
+              Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
+              for (unsigned int i = 0; i < _diag_save_in.size(); ++i)
+              {
+                std::vector<dof_id_type> diag_save_in_dofs(2);
+                diag_save_in_dofs[0] = _current_elem->node_ptr(nd)->dof_number(
+                    _diag_save_in[i]->sys().number(), _diag_save_in[i]->number(), 0);
+                diag_save_in_dofs[1] = _pdmesh.nodePtr(neighbors[dg_neighbors[nb2]])
+                                           ->dof_number(_diag_save_in[i]->sys().number(),
+                                                        _diag_save_in[i]->number(),
+                                                        0);
+
+                _diag_save_in[i]->sys().solution().add_vector(diag, diag_save_in_dofs);
+              }
+            }
+          }
+      }
+  }
+}
+
+void
+HorizonStabilizedFormIIFiniteStrainMechanicsNOSPD::computeLocalOffDiagJacobian(
+    unsigned int jvar_num, unsigned int coupled_component)
+{
+  if (coupled_component == 3) // temperature is coupled
```

Comment:
These special cases of 3 or 4 for coupled_component aren't how we should handle these. For coupling with temperature, we should add a parameter to provide the temperature variable, and check whether jvar_num equals the variable number for temperature. Same deal for the out-of-plane strain. I'm pretty sure if you look at the standard tensor_mechanics code, you can see how to do this.

This shows up elsewhere as well, so please fix all instances of it.